auacm/angular/src/app/judge/judge.component.ts

```diff
@@ -0,0 +1,194 @@
+import { Component, OnInit } from '@angular/core';
+import { Router, ActivatedRoute, Params } from '@angular/router';
+
+import 'rxjs/add/operator/switchMap';
+import { Subject, Observable } from 'rxjs';
+
+import { UserService } from '../user.service';
+import { SubmissionService } from '../submission.service';
+import { ProblemService } from '../problem.service';
+
+import { UserData } from '../models/user';
+import { Problem } from '../models/problem';
+import { RecentSubmission } from '../models/submission';
+
+declare var $:any;
+
+/**
+ * This component is a little crazy. It uses some Angular magic to create searchable problems.
+ */
+
+@Component({
+  selector: 'app-judge',
+  templateUrl: './judge.component.html',
+  styleUrls: ['./judge.component.css']
+})
+export class JudgeComponent implements OnInit {
+
+  user: UserData;
+
+  file: File = undefined;
+
+  searchFilter: string = ""
+  problem: Problem = new Problem();
+  problems: Problem[] = [];
+  problemsMap: Map<number, Problem> = new Map<number, Problem>();
+  problemsObs: Observable<Problem[]>;
+
+  private searchTerms = new Subject<string>();
+
+  highlightIndex = 0;
+
+  submitted: RecentSubmission[] = [];
+
+  constructor(private _userService: UserService,
+              private _submissionService: SubmissionService,
+              private _problemService: ProblemService,
+              private _router: Router, private _route: ActivatedRoute) {
+    this._userService.userData$.subscribe(user => {
+      if (!this.user.loggedIn && user.loggedIn) {
+        this._submissionService.refreshSubmits(user.username, 10);
+      }
+      this.user = user;
+      if (!this.user.loggedIn) {
+        if (this._router.url.startsWith('/judge')) {
+          this._router.navigate(['404']);
+        }
+      }
+    });
+    this._submissionService.recentSubmitsData$.subscribe(submissions => {
+      this.submitted = submissions;
+    })
+  };
+
+  python: any = {
+    version: 'py3'
+  };
+
+  /**
+   * This is needed
+   */
+  search(event: any, term: string) {
+    if (event.keyCode === 40) {
```

Comment:
Can we define constants for 40, 13, and 38 so these keyCodes aren't magic numbers?